spec-0007/index.md

```diff
@@ -0,0 +1,149 @@
+---
+title: "SPEC 7 — Seeding pseudo-random number generation"
+date: 2023-04-19
+author:
+  - "Stéfan van der Walt <stefanv@berkeley.edu>"
+  - "Sebastian Berg <sebastianb@nvidia.com>"
+  - "Pamphile Roy <roy.pamphile@gmail.com>"
+  - "Matt Haberland <mhaberla@calpoly.edu>"
+  - Other participants in the discussion <not.yet@named.org>"
+discussion: https://github.com/scipy/scipy/issues/14322
+endorsed-by:
+---
+
+## Description
+
+Currently, libraries across the ecosystem provide various APIs for seeding random number generation.
+This SPEC suggests a unified, pragmatic API, taking into account technical and historical factors.
+Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.
+
+We recommend:
+
+- avoiding the use of global state and legacy bitstream generators, and
+- standardizing the usage and interpretation of an `rng` keyword for seeding.
+
+We suggest implementing these principles by:
+
+- deprecating the use of `numpy.random.seed` to control the random state,
+- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument, and
+- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.
+
+Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.
+
+### Scope
+
+This is intended as a recommendation to all libraries that allow users to control the
+state of a NumPy random number generator. It is specifically targeted toward functions
+that allow use of `numpy.random.seed` to control the random state and accept
+`RandomState` instances via an argument with a name other than `rng`, but the ideas
+are more broadly applicable. Use of random number generators other than those provided
+by NumPy are beyond the scope of this SPEC.
+
+### Concepts
+
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
+- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
+- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.
+
+## Implementation
+
+Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
+
+Two strong motivations for moving over to `Generator`s are:
```

Comment:
I recommend splitting this off into a separate `## Motivation` section which goes above the Implementation section. That will help the reader, and it will make clearer to the authors how short the motivation given now is (it needs expanding).

Reply:
Can do.

Reply:
Partially addressed by https://github.com/scientific-python/specs/pull/180#discussion_r1705091496, but I leave addition of motivation for a future PR. I'm not the best person to comment on this; I was comfortable controlling using `np.random.rand` and only changed due to peer pressure : )

Reply:
Yup, fixed it.

Reply:
The SPEC template doesn't have a motivation section (/cc @jarrodmillman perhaps we should consider), but I added some motivational text to the description.